library/std/src/io/buffered/bufwriter.rs

```diff
@@ -319,6 +333,95 @@ impl<W: Write> BufWriter<W> {
         let buf = if !self.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
         (self.inner.take().unwrap(), buf)
     }
+
+    // Ensure this function does not get inlined into `write`, so that it
+    // remains inlineable and its common path remains as short as possible.
+    // If this function ends up being called frequently relative to `write`,
+    // it's likely a sign that the client is using an improperly sized buffer
+    // or their write patterns are somewhat pathological.
+    #[cold]
+    fn write_cold(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if buf.len() > self.spare_capacity() {
+            self.flush_buf()?;
+        }
+
+        // Why not len > capacity? To avoid a needless trip through the buffer when the input
+        // exactly fills it. We'd just need to flush it to the underlying writer anyway.
+        if buf.len() >= self.buf.capacity() {
+            self.panicked = true;
+            let r = self.get_mut().write(buf);
+            self.panicked = false;
+            r
+        } else {
+            // Write to the buffer. In this case, we write to the buffer even if it fills it
+            // exactly. Doing otherwise would mean flushing the buffer, then writing this
+            // input to the inner writer, which in many cases would be a worse strategy.
+
+            // SAFETY: We just called `self.flush_buf()`, so `self.buf.len()` is 0, and
+            // we entered this else block because `buf.len() < self.buf.capacity()`.
```

Comment:
> We just called `self.flush_buf()`

That's not true when `buf.len() == self.spare_capacity()` but `self.buf.len() > 0`. (Related: Would it be better to check for that case in `write`, before calling `write_cold`, as it doesn't call flush or the underlying write in that case? I assume you tried this too?)

Reply:
Thanks! Fixed the comment.

I used to have the check in `write` and `write_all`, but I moved it into the cold functions for the sake of improving performance of what I considered to be the common case. That said, I don't have a strong opinion on this, so I can change it if desired.